# Setup Deep Learning Framework

## Daftar Isi
1. Persiapan Environment
2. Instalasi Framework
3. Verifikasi Instalasi
4. Perbandingan Framework
5. Hello World dengan PyTorch
6. Hello World dengan TensorFlow/Keras
7. Tips dan Troubleshooting

## 1. Persiapan Environment

Sebelum menginstal framework deep learning, kita perlu memastikan environment Python kita siap. Mari kita cek versi Python dan package yang terinstal:

In [62]:
import sys
import platform

print(f"Python version: {sys.version}")
print(f"Platform: {platform.platform()}")

# Cek package yang sudah terinstall
import pkg_resources
installed_packages = [dist.project_name for dist in pkg_resources.working_set]
dl_packages = [pkg for pkg in installed_packages if pkg.lower() in ['torch', 'tensorflow', 'keras']]

print("\nInstalled Deep Learning Frameworks:")
if dl_packages:
    for pkg in dl_packages:
        print(f"- {pkg}")
else:
    print("No deep learning frameworks installed yet.")

Python version: 3.13.3 (main, Apr  9 2025, 07:44:25) [GCC 14.2.1 20250207]
Platform: Linux-6.6.87.1-microsoft-standard-WSL2-x86_64-with-glibc2.41

Installed Deep Learning Frameworks:
- keras


/tmp/ipykernel_31959/2099186999.py:8: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
  import pkg_resources


## 2. Instalasi Framework

Untuk menginstal framework deep learning, kita bisa menggunakan pip atau conda. Berikut adalah perintah instalasi untuk masing-masing framework:

### PyTorch
```bash
pip install torch torchvision torchaudio
```

### TensorFlow
```bash
pip install tensorflow
```

### Keras (sekarang bagian dari TensorFlow)
```bash
pip install keras
```

Mari kita coba menginstal framework yang belum terinstal:

## Membuat Virtual Environment

Sebelum menginstal framework, sangat direkomendasikan untuk membuat virtual environment terlebih dahulu. Ini akan membantu mengisolasi package dan dependencies.

1. Buat virtual environment:
```bash
python -m venv dl_env
```

2. Aktifkan virtual environment:
- Windows:
```bash
dl_env\Scripts\activate
```
- Linux/Mac:
```bash
source dl_env/bin/activate
```

3. Install framework dalam virtual environment:
```bash
pip install torch torchvision torchaudio tensorflow
```

Mari kita lakukan langkah-langkah di atas:

## 3. Verifikasi Instalasi

Setelah instalasi selesai, mari kita verifikasi bahwa PyTorch sudah terinstal dengan benar dan bisa digunakan. 

> **Catatan tentang TensorFlow**: 
> - TensorFlow belum mendukung Python 3.13 (versi yang kita gunakan)
> - Untuk menggunakan TensorFlow, disarankan menggunakan Python 3.8-3.11
> - Alternatifnya, kita bisa menggunakan Keras yang sudah terinstal

Mari kita cek instalasi PyTorch:

In [63]:
import torch
import torchvision
import keras

print("PyTorch version:", torch.__version__)
print("CUDA available:", torch.cuda.is_available())
print("Keras version:", keras.__version__)

# Test PyTorch tensor operations
x = torch.rand(5, 3)
print("\nRandom PyTorch tensor:")
print(x)

# Test basic PyTorch operations
y = torch.ones(5, 3)
print("\nAddition of tensors:")
print(x + y)

# Test GPU availability
if torch.cuda.is_available():
    print("\nGPU Information:")
    print("GPU Device:", torch.cuda.get_device_name(0))
else:
    print("\nNo GPU available, using CPU")

ModuleNotFoundError: No module named 'torch'

## Penting: Mengaktifkan Virtual Environment di Jupyter

Setelah membuat virtual environment, kita perlu:
1. Menginstall ipykernel di virtual environment
2. Menambahkan kernel baru untuk Jupyter
3. Memilih kernel baru di notebook ini

Mari kita lakukan langkah-langkah tersebut:

## 4. PyTorch Basics

PyTorch adalah framework deep learning yang populer karena:
1. Sintaks yang Pythonic
2. Dynamic computational graphs
3. Ekosistem yang kuat
4. Debugging yang mudah

Mari kita pelajari konsep dasar PyTorch:

### 4.1 Tensors
Tensor adalah tipe data fundamental di PyTorch, mirip dengan NumPy array tapi bisa dijalankan di GPU.

In [64]:
import torch
import numpy as np

# Membuat tensor dari list
tensor_from_list = torch.tensor([[1, 2, 3], [4, 5, 6]])
print("Tensor dari list:")
print(tensor_from_list)
print("Shape:", tensor_from_list.shape)

# Membuat tensor dari NumPy array
numpy_array = np.array([[1, 2, 3], [4, 5, 6]])
tensor_from_numpy = torch.from_numpy(numpy_array)
print("\nTensor dari NumPy array:")
print(tensor_from_numpy)

# Membuat tensor random
random_tensor = torch.rand(2, 3)
print("\nTensor random:")
print(random_tensor)

# Operasi matematika
print("\nOperasi matematika:")
print("Penjumlahan:", tensor_from_list + 1)
print("Perkalian:", tensor_from_list * 2)
print("Matrix multiplication:\n", torch.mm(tensor_from_list, tensor_from_list.T))

ModuleNotFoundError: No module named 'torch'